src/java.base/share/classes/java/security/PEMEncoder.java

```diff
@@ -0,0 +1,344 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.security;
+
+import jdk.internal.javac.PreviewFeature;
+import sun.security.pkcs.PKCS8Key;
+import sun.security.util.DerOutputStream;
+import sun.security.util.DerValue;
+import sun.security.util.Pem;
+import sun.security.x509.AlgorithmId;
+
+import javax.crypto.*;
+import javax.crypto.spec.PBEKeySpec;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.*;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Objects;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * PEMEncoder is an immutable class for Privacy-Enhanced Mail (PEM)
+ * data.  PEM is a textual encoding used to store and transfer security
+ * objects, such as asymmetric keys, certificates, and certificate revocation
+ * lists (CRL).  It is defined in RFC 1421 and RFC 7468.  PEM consists of a
+ * Base64-formatted binary encoding enclosed by a type-identifying header
+ * and footer.
+ *
+ * <p> Encoding may be performed on Java Cryptographic Extension (JCE) objects
+ * that implement {@link DEREncodable} and support
+ * {@linkplain PKCS8EncodedKeySpec PKCS#8} or
+ * {@linkplain X509EncodedKeySpec X509} formats.
+ *
+ * <p> Encrypted private key PEM data can be built by calling the encode methods
+ * on a PEMEncoder instance returned by {@link #withEncryption(char[])} or
+ * by passing an {@link EncryptedPrivateKeyInfo} object into the encode methods.
+ *
+ * <p> PKCS8 2.0 allows OneAsymmetricKey encoding, which may contain both private
+ * and public keys in the same PEM.This is supported by using the
+ * {@link KeyPair} class with the encode methods.
+ *
+ * <p> When encoding a {@link PEMRecord}, the API surrounds the
+ * {@linkplain PEMRecord#pem()} with a generated the PEM header and footer
+ * from {@linkplain PEMRecord#type()}.  It will not check the validity of
+ * the data.
+ *
+ * <p>{@code String} values returned by this class use character set
+ * {@link java.nio.charset.StandardCharsets#ISO_8859_1 ISO-8859-1}.
+ *
+ * @apiNote
+ * Here is an example of encoding a PrivateKey object:
+ * {@snippet lang = java:
+ *     PEMEncoder pe = PEMEncoder.of();
+ *     byte[] pemData = pe.encode(privKey);
+ * }
+ *
+ * @see PKCS8EncodedKeySpec
+ * @see X509EncodedKeySpec
```

Comment:
I cannot see how these 2 deserve this place. I'd rather link to `PEMRecord` and `PEMDecoder`.